File: web/src/api/actions.ts
import { ApiError, api } from "./client";
import type { ActionType, ProposeResponse } from "./types";

export interface ActionRequestBody {
  action: ActionType;
  message_ids?: string[];
  search?: string;
  category?: string;
  sender?: string;
  after?: string;
  before?: string;
  unread_only?: boolean;
  limit?: number;
  label_name?: string;
}

export interface RunActionOptions {
  /** archive/trash always confirm regardless of target count, matching the backend/CLI rule. */
  alwaysConfirm?: boolean;
  /** Shows the affected emails and resolves true/false based on the user's choice. */
  confirm: (proposal: ProposeResponse) => Promise<boolean>;
}

/**
 * Single explicit id + a safe action -> executes immediately. Anything else
 * (bulk, search-resolved, or archive/trash) -> propose, show the affected
 * emails via opts.confirm, and only confirm if the user says yes. Mirrors
 * cli/mailchef_cli/actions.py's run_action so the CLI and web UI enforce the
 * same confirmation guarantees.
 */
export async function runAction(body: ActionRequestBody, opts: RunActionOptions): Promise<void> {
  const singleExplicit = !body.search && body.message_ids?.length === 1;

  if (!opts.alwaysConfirm && singleExplicit) {
    try {
      await api.actionsExecute(body as unknown as Record<string, unknown>);
      return;
    } catch (e) {
      if (!(e instanceof ApiError) || e.status !== 400) throw e;
      // Backend says this needs confirmation after all — fall through to propose/confirm.
    }
  }

  const proposal = await api.actionsPropose(body as unknown as Record<string, unknown>);
  if (proposal.affected_count === 0) return;

  const confirmed = await opts.confirm(proposal);
  if (confirmed) {
    await api.actionsConfirm(proposal.proposal_id);
  } else {
    await api.actionsCancel(proposal.proposal_id);
  }
}

File: cli/mailchef_cli/actions.py
from rich.console import Console
from rich.prompt import Confirm
from rich.table import Table

from mailchef_cli.client import ClientError, MailChefClient


def run_action(
    client: MailChefClient,
    console: Console,
    action: str,
    ids: list[str] | None,
    search: str | None,
    category: str | None = None,
    sender: str | None = None,
    after: str | None = None,
    before: str | None = None,
    unread: bool = False,
    limit: int = 100,
    label: str | None = None,
    always_confirm: bool = False,
) -> None:
    """Single explicit id + a safe action -> execute immediately. Anything
    else (bulk, search-resolved, or archive/trash) -> propose, show the
    affected emails, and only confirm if the user explicitly says yes.
    """
    body = {
        "action": action,
        "message_ids": ids or None,
        "search": search,
        "category": category,
        "sender": sender,
        "after": after,
        "before": before,
        "unread_only": unread,
        "limit": limit,
        "label_name": label,
    }

    if not always_confirm and ids and len(ids) == 1 and not search:
        try:
            result = client.post("/actions/execute", body)
            console.print(f"[green]Done:[/green] {action} applied to {result['message_count']} email(s).")
            return
        except ClientError as e:
            if e.status_code != 400:
                raise
            # Backend says this needs confirmation after all (e.g. archive) — fall through.

    proposal = client.post("/actions/propose", body)
    count = proposal["affected_count"]
    if count == 0:
        console.print("[yellow]No matching emails found — nothing to do.[/yellow]")
        return

    console.print(f"\nThis will [bold]{action}[/bold] {count} email(s):\n")
    console.print(_affected_table(proposal["affected"]))

    if not Confirm.ask(f"\nProceed with {action} on {count} email(s)?", default=False):
        client.post("/actions/cancel", {"proposal_id": proposal["proposal_id"]})
        console.print("[yellow]Cancelled.[/yellow]")
        return

    result = client.post("/actions/confirm", {"proposal_id": proposal["proposal_id"]})
    console.print(f"[green]Done:[/green] {action} applied to {result['message_count']} email(s).")


def _affected_table(affected: list[dict]) -> Table:
    table = Table(show_header=True, header_style="bold")
    table.add_column("From")
    table.add_column("Subject")
    table.add_column("Date")
    table.add_column("Category")
    for m in affected:
        table.add_row(m["from"], m["subject"], (m.get("date") or "")[:10], m.get("category") or "")
    return table
